### ISSUE TITLE
Occasional failures to start new orchestrations

### ISSUE THREAD

**Issue by fabioenc00** [microsoft/durabletask-netherite#223]
Hi,

We're getting some exceptions every day when starting new orchestrations. StartNewAsync method seems to fail to start a new orchestration when we have a higher volume of requests. We can't seem to figure out the reason why this is happening as this exception is occuring in DurableTask.Netherite package.

Could you take a look please?

Thank you!

Code:
```
...

string instanceId = await starter.StartNewAsync(_orchestratorName, input);

...
```
_orchestratorName is a constant variable with the name of the orchestrator.

Exception: NullReferenceException at DurableTask.Netherite.NetheriteOrchestrationService

Stack trace:

```
Microsoft.Azure.WebJobs.Host.FunctionInvocationException:
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithLoggingAsync>d__26.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:352)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<TryExecuteAsync>d__18.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:108)
Inner exception System.NullReferenceException handled at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw:
   at DurableTask.Netherite.NetheriteOrchestrationService+<DurableTask-Core-IOrchestrationServiceClient-CreateTaskOrchestrationAsync>d__98.MoveNext (DurableTask.Netherite, Version=1.0.0.0, Culture=neutral, PublicKeyToken=ef8c4135b1b4225a)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at DurableTask.Core.TaskHubClient+<InternalCreateOrchestrationInstanceWithRaisedEventAsync>d__27.MoveNext (DurableTask.Core, Version=2.11.0.0, Culture=neutral, PublicKeyToken=d53979610a6e89dd: /_/src/DurableTask.Core/TaskHubClient.cs:614)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableClient+<Microsoft-Azure-WebJobs-Extensions-DurableTask-IDurableOrchestrationClient-StartNewAsync>d__34`1.MoveNext (Microsoft.Azure.WebJobs.Extensions.DurableTask, Version=2.0.0.0, Culture=neutral, PublicKeyToken=014045d636e89289: D:\a\_work\1\s\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableClient.cs:210)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at My.Functions.Clients.FunctionStartOrchestration+<RunAndWait>d__4.MoveNext (My.Functions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null: C:\Projetos\AzureFunctions\MyAzureFunctionsProject\My.Functions\My.Functions\Clients\FunctionStartOrchestration.cs:64)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2+<InvokeAsync>d__10.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:52)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<InvokeWithTimeoutAsync>d__33.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:581)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithWatchersAsync>d__32.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:527)
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult (System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e)
   at Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor+<ExecuteWithLoggingAsync>d__26.MoveNext (Microsoft.Azure.WebJobs.Host, Version=3.0.34.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35: C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs:306)

```

**Comment by sebastianburckhardt** [comment]
Based on the stack trace, my best guess is that there is a race condition between the application code trying to use the client and a host shutdown in progress that is stopping everything, including the client. In this situation, the code is throwing NullReferenceException.

I can fix this by changing the code so it throws a more obvious exception, for example an OperationCancelledException with a message saying "could not start orchestration because the host is shutting down". You would still need to handle the exception, I don't see any way of completely preventing this situation (when the client is shut down it can no longer be used to start an orchestration).

If possible, I would like to confirm that I am interpreting the situation correctly. Did you run this in a cloud deployment? If so I may be able to take a look at the traces to confirm my suspicion. In that case I would need the app name and an approximate timestamp for when you observed the exception.

**Comment by fabioenc00** [comment]
Hi Sebastian,

Yes, this exception is happening in our production environment in Azure.

We're running EP2 plan, 3 always ready instances and no scale out limit.

CPU usage is low (average under 30%) as well as memory usage.

Reasonable amount of function executions:
(Function Execution Count per minute)
![image](https://user-images.githubusercontent.com/123659141/215894665-526648ba-fa6f-40df-ae68-5646ccffe9cd.png)



This was today, we had 36 of this NullReferenceException in the last 24 hours.

(Time is in UTC in the screenshot)
![image](https://user-images.githubusercontent.com/123659141/215890967-3976b94b-a258-47cd-b267-d1ad98f8776e.png)

We have also opened an Azure Support ticket (ID: 2301240050002113) a week ago and have been in contact every day since with MS Support to try to identify and solve this issue, but so far it had not been possible yet to identify the reason behind this exceptions or a solution, so we really appreciate your help.

> In that case I would need the app name and an approximate timestamp for when you observed the exception.

We're glad to provide you this. You can see the timestamps on the app insights screenshot, about the app name is it possible to share it with you in a private channel? (maybe email or other means if possible).

Thank you once again.

**Comment by sebastianburckhardt** [comment]
Thanks for the information. I was able to confirm that all of the error messages in the above log are happening at the same time as a host shutdown, so it is very likely that they are indeed caused by the race condition I suspected.

Now, as to what to do about it:
- I will fix the code to throw a more meaningful exception (OperationCanceledException)
- I will reduce the window for the race condition so exceptions are less likely (i.e. keep the client working for as long as possible)

However, I cannot guarantee that no exceptions are ever thrown. Because I think it will remain possible to run into timings where a client request races with a host shutdown and can no longer proceed successfully. In such situations the application will need to deal with the OperationCanceledException, for example by aborting the request in a way so it can be retried on a different host. 

That would also be my recommended workaround for the time being, until my fix is in place.

**Comment by fabioenc00** [comment]
Great, thanks for the quick response!

>  In such situations the application will need to deal with the OperationCanceledException, for example by aborting the request in a way so it can be retried on a different host. 
That would also be my recommended workaround for the time being, until my fix is in place.

Would it be possible for you to give us an example or an idea on how we could do this? Aborting the request and retrying it on a different host.

Much appreciated.

Thanks!

**Comment by sebastianburckhardt** [comment]
> Would it be possible for you to give us an example or an idea on how we could do this? Aborting the request and retrying it on a different host.

I haven't tried this myself but here are some ideas on how I would approach this problem, i.e. what I think is the "most logical way" to handle this situation. What to do depends on how the code that calls the client is triggered in the first place. 

- For queue or event hubs triggers, it is my understanding that this should already work by default. Those triggers perform retries when an exception is thrown; and if the host is shutting down, the retry loop should exit and instead leave the message in the queue, to be handled by a different host next time. 

- In an Http trigger, I would return a special http error response (e.g. 503 Server unavailable) which should then in turn cause the client to reissue the request (it may be necessary to modify the client code so it retries on 503 responses if it does not already do that), and then this retried request would get routed to a different host that is not being shutdown and can handle the request (I think the load balancer would no longer send requests to a host that has already initiated shutdown)

**Comment by fabioenc00** [comment]
Hi @sebastianburckhardt,

I see that you closed this issue last week, does it mean that you have completed these changes?

> - I will fix the code to throw a more meaningful exception (OperationCanceledException)
> - I will reduce the window for the race condition so exceptions are less likely (i.e. keep the client working for as long as possible)

If yes, could you please tell me just an estimate of when you are planning on releasing them?

We have implemented the workaround you suggested for the time being and we could mitigate almost all of the issues, but we still have some occurrences every now and then, but I think we will stop having issues at all with this changes you mentioned.

Thanks!!

**Comment by sebastianburckhardt** [comment]
Apologies, I closed this too early. I somehow thought the pull request was already merged but it's actually still sitting in my local repo.

**Comment by sebastianburckhardt** [comment]
Fix was released with 1.3.3.

Let me know if this works for you.

Based on my own tests, I am not sure whether these exceptions are thrown *less* often now. They may actually be thrown *more* often since the implementation is now more diligent about telling the callers that the client is shutting down (as opposed to just never returning the call at all).
